sentry_sdk/integrations/rust_tracing.py

```diff
@@ -0,0 +1,259 @@
+
+Using it requires additional setup on the Rust side to accept a
+`RustTracingLayer` Python object and register it with the `tracing-subscriber`
+using an adapter from the `pyo3-python-tracing-subscriber` crate. For example:
+```rust
+#[pyfunction]
+pub fn initialize_tracing(py_impl: Bound<'_, PyAny>) {
+    tracing_subscriber::registry()
+        .with(pyo3_python_tracing_subscriber::PythonCallbackLayerBridge::new(py_impl))
+        .init();
+}
+```
+
+Usage in Python would then look like:
+```
+sentry_sdk.init(
+    dsn=sentry_dsn,
+    integrations=[
+        SentryIntegrationFactory.create(
+            "demo_rust_extension",
+            demo_rust_extension.initialize_tracing,
+            event_type_mapping=event_type_mapping,
+        )
+    ],
+)
+```
+
+Each native extension requires its own integration.
+"""
+
+import json
+from enum import Enum, auto
+from typing import Any, Callable, Final, Tuple, Optional
+
+import sentry_sdk
+from sentry_sdk.integrations import Integration
+from sentry_sdk.tracing import Span as SentrySpan
+
+TraceState = Optional[Tuple[SentrySpan | None, SentrySpan]]
+
+
+class RustTracingLevel(Enum):
+    Trace: Final[str] = "TRACE"
+    Debug: Final[str] = "DEBUG"
+    Info: Final[str] = "INFO"
+    Warn: Final[str] = "WARN"
+    Error: Final[str] = "ERROR"
+
+
+class EventTypeMapping(Enum):
+    Ignore = auto()
+    Exc = auto()
+    Breadcrumb = auto()
+    Event = auto()
+
+
+def tracing_level_to_sentry_level(level):
+    # type: (str) -> sentry_sdk._types.LogLevelStr
+    match RustTracingLevel(level):
+        case RustTracingLevel.Trace | RustTracingLevel.Debug:
+            return "debug"
+        case RustTracingLevel.Info:
+            return "info"
+        case RustTracingLevel.Warn:
+            return "warning"
+        case RustTracingLevel.Error:
+            return "error"
+
+
+def extract_contexts(event: dict[str, Any]) -> dict[str, Any]:
+    metadata = event.get("metadata", {})
+    contexts = {}
+
+    location = {}
+    for field in ["module_path", "file", "line"]:
+        if field in metadata:
+            location[field] = metadata[field]
+    if len(location) > 0:
+        contexts["Rust Tracing Location"] = location
+
+    fields = {}
+    for field in metadata.get("fields", []):
+        fields[field] = event.get(field)
+    if len(fields) > 0:
+        contexts["Rust Tracing Fields"] = fields
+
+    return contexts
+
+
+def process_event(event: dict[str, Any]) -> None:
+    metadata = event.get("metadata", {})
+
+    logger = metadata.get("target")
+    level = tracing_level_to_sentry_level(metadata.get("level"))
+    message = event.get("message")  # type: sentry_sdk._types.Any
+    contexts = extract_contexts(event)
+
+    sentry_event = {
+        "logger": logger,
+        "level": level,
+        "message": message,
+        "contexts": contexts,
+    }  # type: sentry_sdk._types.Event
+
+    sentry_sdk.capture_event(sentry_event)
```

Comment:
this is a draft PR because this line is giving me some grief

in my example project, each span emits a debug-level event or two. if `event_type_mapping` is used to ignore them, the "naive fibonacci" transaction and its whole span tree show up like i'd expect in the "performance" and "traces" tabs. however, if the events are allowed to be logged, the transaction + span tree are lost and all i get are issues
<img width="1278" alt="Screenshot 2024-10-29 at 7 52 17 PM" src="https://github.com/user-attachments/assets/710d1cdc-35a7-4255-b68d-598705dc2e35">

if i click through one of the issues and look at the trace it came from, i see nonsense:
<img width="1238" alt="Screenshot 2024-10-29 at 7 53 14 PM" src="https://github.com/user-attachments/assets/d4a8a28d-a190-4c22-a389-ce3572acb224">

notice that the events know they come from the "naive fibonacci" transaction, but the transaction otherwise does not show up in the UI and is not useful

Reply:
* is the transaction sent at all from the SDK in the latter case? check with `sentry_sdk.init(debug=True)`
* what is the lifecycle like?  
* what is the actual behaviour you want for those debug-level events? do you want them to create issues or not?

when you say 
> each span emits a debug-level event 

is the span always going through `on_new_span` and those debug-level events through `on_event`?

Reply:
I think it might be wiser to map the debug events to breadcrumbs rather than making issues out of all of them.

Reply:
> is the transaction sent at all from the SDK in the latter case? check with sentry_sdk.init(debug=True)

i poked around with this and discovered i am simply logging too many events. thanks!

showing my work: "naive fibonacci" is a 2**n algorithm that should generate 2**n debug logs. i called it with n=10, so there should have been 2**10 events. but the transport queue only ever had suspiciously close to 100 events in it. that pointed me to
https://github.com/getsentry/sentry-python/blob/02d09346e6d070e03b828807d72485b6f23b2c11/sentry_sdk/consts.py#L80

so my transaction envelope was getting pushed off the back of the truck. if i call with n=6 instead, everything fits in the transport queue and arrives to sentry

> what is the lifecycle like? is the span always going through on_new_span and those debug-level events through on_event?

yes

> what is the actual behaviour you want for those debug-level events? do you want them to create issues or not? I think it might be wiser to map the debug events to breadcrumbs rather than making issues out of all of them.

the purpose of the project i am asking about was to demo all of the behavior of the integration:
- transaction/span tree, extra fields set on spans
- breadcrumbs
- issues/events, contexts set on them

so yes, a real-world project would want breadcrumbs instead (or nothing) for these particular logs. by default, the integration maps warn/info to breadcrumbs and completely ignores debug/trace, i just overrode that for sake of testing/demoing for this project